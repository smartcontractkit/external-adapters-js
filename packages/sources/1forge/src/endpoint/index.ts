import type { TInputParameters as ConvertInputParameters } from './convert'
import type { TInputParameters as QuotesInputParameters } from './quotes'

export type TInputParameters = ConvertInputParameters | QuotesInputParameters

export * as quotes from './quotes'
export * as convert from './convert'
