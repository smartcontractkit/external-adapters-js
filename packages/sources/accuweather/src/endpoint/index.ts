export * as currentConditions from './current-conditions'
export * as locationCurrentConditions from './location-current-conditions'
export * as location from './location'
