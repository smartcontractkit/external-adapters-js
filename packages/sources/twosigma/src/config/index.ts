import { AdapterConfig } from '@chainlink/external-adapter-framework/config'

export const config = new AdapterConfig({
  WS_API_ENDPOINT: {
    description:
      'The WebSocket API URL. Either "wss://chainlinkcloud1.twosigma.com:8765" or "wss://chainlinkcloud1.twosigma.com:8766"',
    type: 'string',
    required: true,
    default: 'wss://chainlinkcloud1.twosigma.com:8765',
  },
  WS_API_KEY: {
    description: 'The API key used to authenticate requests',
    type: 'string',
    required: true,
    sensitive: true,
  },
})
