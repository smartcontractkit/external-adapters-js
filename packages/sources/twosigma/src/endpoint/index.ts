export { endpoint as priceEndpoint } from './price'
