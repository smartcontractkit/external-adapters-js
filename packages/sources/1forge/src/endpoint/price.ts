import { Requester, Validator } from '@chainlink/ea-bootstrap'
import { ExecuteWithConfig, Config } from '@chainlink/types'
import { NAME as AdapterName } from '../config'

export const NAME = 'price'

const customParams = {
  base: ['base', 'from'],
  quote: ['quote', 'to'],
  quantity: false,
}

export const execute: ExecuteWithConfig<Config> = async (request, config) => {
  const validator = new Validator(request, customParams)
  if (validator.error) throw validator.error

  const jobRunID = validator.validated.id
  const url = `/convert`
  const from = (validator.overrideSymbol(AdapterName) as string).toUpperCase()
  const to = validator.validated.data.quote.toUpperCase()
  const quantity = validator.validated.data.quantity || 1

  const params = {
    ...config.api.params,
    from,
    to,
    quantity,
  }

  const options = {
    ...config.api,
    url,
    params,
  }

  const response = await Requester.request(options)
  response.data.result = Requester.validateResultNumber(response.data, ['value'])

  return Requester.success(jobRunID, response, config.verbose)
}
