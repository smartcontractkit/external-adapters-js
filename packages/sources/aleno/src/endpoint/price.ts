import { AdapterEndpoint } from '@chainlink/external-adapter-framework/adapter'
import { InputParameters } from '@chainlink/external-adapter-framework/validation'
import { SingleNumberResultResponse } from '@chainlink/external-adapter-framework/util'
import { TransportRoutes } from '@chainlink/external-adapter-framework/transports'
import { priceEndpointInputParametersDefinition } from '@chainlink/external-adapter-framework/adapter'
import { config } from '../config'

import { httpTransport } from '../transport/price-http'
import { socketioTransport } from '../transport/price-socketio'

export const inputParameters = new InputParameters(priceEndpointInputParametersDefinition, [
  {
    base: 'FRAX',
    quote: 'USD',
  },
])

export type BaseEndpointTypes = {
  Parameters: typeof inputParameters.definition
  Response: SingleNumberResultResponse
  Settings: typeof config.settings
}

export const endpoint = new AdapterEndpoint({
  name: 'price',
  aliases: ['crypto', 'state'],
  transportRoutes: new TransportRoutes<BaseEndpointTypes>()
    .register('rest', httpTransport)
    .register('socketio', socketioTransport),
  inputParameters,
  defaultTransport: 'socketio',
})
