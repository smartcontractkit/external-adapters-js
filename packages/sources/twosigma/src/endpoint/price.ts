import {
  EndpointContext,
  PriceEndpoint,
  priceEndpointInputParametersDefinition,
} from '@chainlink/external-adapter-framework/adapter'
import { WebSocketTransport } from '@chainlink/external-adapter-framework/transports'
import { SubscriptionDeltas } from '@chainlink/external-adapter-framework/transports/abstract/streaming'
import { makeLogger, ProviderResult } from '@chainlink/external-adapter-framework/util'

import { config } from '../config'
import { InputParameters } from '@chainlink/external-adapter-framework/validation'

const inputParameters = new InputParameters(priceEndpointInputParametersDefinition)
export type RequestParams = typeof inputParameters.validated

// Schema of message sent to Two Sigma to start streaming symbols
export type WebSocketRequest = {
  api_key: string
  symbols: string[] // e.g. ["AAPL/USD"]
}

// Schema of message sent from Two Sigma containing symbol prices
export type WebSocketMessage = {
  timestamp: number // The UTC timestamp in seconds. e.g. 1666705053.713266
  symbol_price_dict: Record<
    string, // The ticker symbol. e.g. AAPL/USD
    SymbolPriceData
  >
}

export type SymbolPriceData = {
  quote_currency: string // Quote currency. e.g. USD
  session_status_flag:
    | 'premarket' // Pre-market trading during pre-market hours (before 9.30am)
    | 'open' // Market is open and trading during market hours
    | 'postmarket' // After-hours trading during post-market hours
    | 'closed' // Market is closed
  asset_status_flag:
    | 'active' // TSS is trading this stock actively
    | 'inactive' // TSS has not traded this stock recently
  confidence_interval: number // Confidence interval in quote currency units. e.g. 0.16416514635149188
  price: number // Price in quote currency units. e.g. 379.64
}

export type WebSocketEndpointTypes = {
  // i.e. { base, quote }
  // base is the symbol to query, e.g. AAPL
  // quote is the currency to convert to, e.g. USD
  Parameters: typeof inputParameters.definition
  Response: {
    Data: {
      result: number
    }
    Result: number
  }
  Settings: typeof config.settings
  Provider: {
    WsMessage: WebSocketMessage
  }
}

const logger = makeLogger('TwoSigmaPriceWebsocketEndpoint')

export class TwoSigmaWebsocketTransport extends WebSocketTransport<WebSocketEndpointTypes> {
  // The API works by first sending a message to the server containing the API key
  // and a list of ticker symbols to subscribe to, after which the server begins
  // streaming price updates. However, in order to subscribe to a different set of
  // symbols, the client would need to do so on a new websocket connection as the
  // server seems to ignore all messages after the first.

  override async streamHandler(
    context: EndpointContext<WebSocketEndpointTypes>,
    subscriptions: SubscriptionDeltas<RequestParams>,
  ): Promise<void> {
    if (
      this.wsConnection &&
      !this.connectionClosed() &&
      (subscriptions.new.length || subscriptions.stale.length)
    ) {
      logger.debug(
        `closing WS connection for new subscriptions: ${JSON.stringify(subscriptions.desired)}`,
      )

      const closed = new Promise<void>((resolve) => (this.wsConnection.onclose = resolve))
      this.wsConnection.close()
      await closed
    }

    subscriptions.new = subscriptions.desired
    subscriptions.stale = []

    await super.streamHandler(context, subscriptions)
  }

  override async sendMessages(
    _context: EndpointContext<WebSocketEndpointTypes>,
    subscribes: RequestParams[],
    unsubscribes: unknown[],
  ): Promise<void> {
    // Send a single message containing the entire set of subscribed symbols rather than
    // one message per symbol.

    if (unsubscribes.length) {
      // We explicitly set subscriptions.stale to empty.
      logger.warn(`unexpected unsubscribes: ${JSON.stringify(unsubscribes)}`)
    }

    if (!subscribes.length) {
      // No symbols to subscribe to
      logger.debug(`nothing to subscribe to, skipping initial message`)
      return
    }

    const payload: WebSocketRequest = {
      api_key: process.env.WS_API_KEY || 'twosigma api key not set',
      symbols: subscribes.map(buildSymbol).sort(),
    }
    this.wsConnection.send(this.serializeMessage(payload))
  }
}

export const parseBaseQuote = (symbol: string): RequestParams | undefined => {
  // "AAPL/USD" -> { base: AAPL, quote: USD }
  const splits = symbol.split('/')
  if (splits.length !== 2) {
    return
  }
  const [base, quote] = splits
  return { base, quote }
}

export const buildSymbol = ({ base, quote }: RequestParams): string => {
  return `${base}/${quote}`
}

export const options = {
  url: (context: EndpointContext<WebSocketEndpointTypes>): string => {
    return context.adapterSettings.WS_API_ENDPOINT
  },

  handlers: {
    message: (message: WebSocketMessage): ProviderResult<WebSocketEndpointTypes>[] | undefined => {
      if (!message.timestamp || !message.symbol_price_dict) {
        logger.debug(`Received invalid message: ${message}`)
        return undefined
      }

      const results: ProviderResult<WebSocketEndpointTypes>[] = []
      for (const symbol in message.symbol_price_dict) {
        const priceData = message.symbol_price_dict[symbol]
        const params = parseBaseQuote(symbol)
        if (params === undefined) {
          continue
        }

        results.push({
          params,
          response: {
            result: priceData.price,
            data: {
              result: priceData.price,
            },
            timestamps: {
              providerIndicatedTimeUnixMs: Math.floor(message.timestamp * 1000),
            },
          },
        })
      }

      return results
    },
  },

  // We don't want the 'subscribeMessage' and 'unsubscribeMessage' hooks because we override
  // the 'sendMessages' on the transport to send a single message for the entire set of
  // symbols. However, this empty object is necessary for the transport to actually call our
  // overridden 'sendMessages'.
  builders: {},
}

export const endpoint = new PriceEndpoint({
  name: 'price',
  aliases: ['stock'],
  inputParameters,
  transport: new TwoSigmaWebsocketTransport(options),
})
