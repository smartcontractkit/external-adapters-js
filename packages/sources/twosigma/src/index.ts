import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { PriceAdapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { priceEndpoint } from './endpoint'

const makeAdapter = () => {
  // Factory method because the adapter loads configs from env vars at initialization
  // time, so this is useful in tests for example.
  return new PriceAdapter({
    name: 'TWOSIGMA',
    endpoints: [priceEndpoint],
    defaultEndpoint: priceEndpoint.name,
    config,
  })
}

export const adapter = makeAdapter()

export const server = (): Promise<ServerInstance | undefined> => expose(makeAdapter())
