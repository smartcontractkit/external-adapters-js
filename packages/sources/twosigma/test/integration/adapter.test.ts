import { AddressInfo } from 'net'
import * as process from 'process'

import { ServerInstance } from '@chainlink/external-adapter-framework'
import { WebSocketClassProvider } from '@chainlink/external-adapter-framework/transports'
import { sleep } from '@chainlink/external-adapter-framework/util'
import { Server, WebSocket } from 'mock-socket'
import request, { SuperTest, Test } from 'supertest'

import { server } from '../../src'
import { WebSocketMessage, WebSocketRequest } from '../../src/endpoint/price'

/**
 * Sets the mocked websocket instance in the provided provider class.
 * We need this here, because the tests will connect using their instance of WebSocketClassProvider;
 * fetching from this library to the \@chainlink/ea-bootstrap package would access _another_ instance
 * of the same constructor. Although it should be a singleton, dependencies are different so that
 * means that the static classes themselves are also different.
 *
 * @param provider - singleton WebSocketClassProvider
 */
export const mockWebSocketProvider = (provider: typeof WebSocketClassProvider): void => {
  // Extend mock WebSocket class to bypass protocol headers error
  class MockWebSocket extends WebSocket {
    constructor(url: string, protocol: string | string[] | Record<string, string> | undefined) {
      super(url, protocol instanceof Object ? undefined : protocol)
    }
    // This is part of the 'ws' node library but not the common interface, but it's used in our WS transport
    removeAllListeners() {
      for (const eventType in this.listeners) {
        // We have to manually check because the mock-socket library shares this instance, and adds the server listeners to the same obj
        if (!eventType.startsWith('server')) {
          delete this.listeners[eventType]
        }
      }
    }
  }

  // Need to disable typing, the mock-socket impl does not implement the ws interface fully
  provider.set(MockWebSocket as any) // eslint-disable-line @typescript-eslint/no-explicit-any
}

describe('websocket', () => {
  const oldEnv = { ...process.env }
  const webSocketEndpoint = 'ws://localhost:9090'
  const webSocketApiKey = 'abcdef'

  let spy: jest.SpyInstance
  let mockWebSockerServer: Server
  let fastify: ServerInstance | undefined
  let req: SuperTest<Test>

  const makeRequest = (base: string) =>
    req
      .post('/')
      .send({ data: { base, quote: 'USD' } })
      .set('Accept', '*/*')
      .set('Content-Type', 'application/json')
      .expect('Content-Type', /json/)

  jest.setTimeout(10000)

  beforeAll(async () => {
    process.env['WS_SUBSCRIPTION_TTL'] = '5000'
    process.env['CACHE_MAX_AGE'] = '5000'
    process.env['CACHE_POLLING_MAX_RETRIES'] = '0'
    process.env['METRICS_ENABLED'] = 'false'
    process.env['WS_API_ENDPOINT'] = webSocketEndpoint
    process.env['WS_API_KEY'] = webSocketApiKey

    const mockDate = new Date('2023-01-01T00:00:00Z')
    spy = jest.spyOn(Date, 'now').mockReturnValue(mockDate.getTime())

    // Initialize adapter before mocking websocket provider
    fastify = await server()
    req = request(`http://localhost:${(fastify?.server.address() as AddressInfo).port}`)

    mockWebSocketProvider(WebSocketClassProvider)
    mockWebSockerServer = new Server(webSocketEndpoint, { mock: false })
    mockWebSockerServer.on('connection', (socket) => {
      socket.on('message', (message) => {
        const payload: WebSocketRequest = JSON.parse(message as string)
        expect(payload.api_key).toEqual(webSocketApiKey)

        const symbolPriceDict: WebSocketMessage['symbol_price_dict'] = {}
        payload.symbols.forEach((symbol, i) => {
          symbolPriceDict[symbol] = {
            quote_currency: 'USD',
            session_status_flag: 'open',
            asset_status_flag: 'active',
            confidence_interval: (i + 1) * 0.1,
            price: (i + 1) * 100,
          }
        })

        socket.send(
          JSON.stringify({
            timestamp: 1645203822,
            symbol_price_dict: symbolPriceDict,
          }),
        )
      })
    })

    // Send initial request to start background execute
    await makeRequest('AAPL')
    await sleep(5000)
  })

  afterAll(async () => {
    spy.mockRestore()
    for (const key in oldEnv) {
      process.env[key] = oldEnv[key]
    }
    await fastify?.close()
    mockWebSockerServer.close()
  })

  describe('price endpoint', () => {
    it('should return success', async () => {
      const resp1 = await makeRequest('AAPL')
      expect(resp1.body).toEqual({
        result: 100,
        data: {
          result: 100,
        },
        timestamps: {
          providerDataReceivedUnixMs: 1672531200000, // mocked time
          providerDataStreamEstablishedUnixMs: 1672531200000,
          providerIndicatedTimeUnixMs: 1645203822000, // response.timestamp
        },
        statusCode: 200,
      })

      await makeRequest('AMZN')
      await sleep(5000) // Wait for next background executor loop

      const resp2 = await makeRequest('AAPL')
      expect(resp2.body).toEqual({
        result: 100,
        data: {
          result: 100,
        },
        timestamps: {
          providerDataReceivedUnixMs: 1672531200000,
          providerDataStreamEstablishedUnixMs: 1672531200000,
          providerIndicatedTimeUnixMs: 1645203822000,
        },
        statusCode: 200,
      })
      const resp3 = await makeRequest('AMZN')
      expect(resp3.body).toEqual({
        result: 200,
        data: {
          result: 200,
        },
        timestamps: {
          providerDataReceivedUnixMs: 1672531200000,
          providerDataStreamEstablishedUnixMs: 1672531200000,
          providerIndicatedTimeUnixMs: 1645203822000,
        },
        statusCode: 200,
      })
    }, 30000)
  })
})
