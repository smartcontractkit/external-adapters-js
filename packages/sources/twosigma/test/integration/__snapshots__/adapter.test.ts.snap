// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`websocket price endpoint should return success 1`] = `
{
  "data": {
    "result": 100,
  },
  "result": 100,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1018,
    "providerDataStreamEstablishedUnixMs": 1010,
    "providerIndicatedTimeUnixMs": 1645203822000,
  },
}
`;

exports[`websocket price endpoint should return success 2`] = `
{
  "data": {
    "result": 200,
  },
  "result": 200,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 2036,
    "providerDataStreamEstablishedUnixMs": 2028,
    "providerIndicatedTimeUnixMs": 1645203822000,
  },
}
`;

exports[`websocket price endpoint should return success 3`] = `
{
  "data": {
    "result": 100,
  },
  "result": 100,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 2036,
    "providerDataStreamEstablishedUnixMs": 2028,
    "providerIndicatedTimeUnixMs": 1645203822000,
  },
}
`;

exports[`websocket price endpoint should return success 4`] = `
{
  "data": {
    "result": 200,
  },
  "result": 200,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 2036,
    "providerDataStreamEstablishedUnixMs": 2028,
    "providerIndicatedTimeUnixMs": 1645203822000,
  },
}
`;
